rcl/test/rcl/test_get_type_description_service.cpp

```diff
@@ -0,0 +1,318 @@
+      GET_TYPE_DESCRIPTION_SRV_TYPE_NAME));
+  EXPECT_EQ(RCL_RET_NOT_INIT, rcl_node_type_description_service_fini(this->node_ptr));
+
+  EXPECT_EQ(RCL_RET_OK, rcl_node_type_description_service_init(this->node_ptr));
+  EXPECT_TRUE(
+    service_exists(
+      this->node_ptr, this->get_type_description_service_name,
+      GET_TYPE_DESCRIPTION_SRV_TYPE_NAME));
+  EXPECT_EQ(RCL_RET_ALREADY_INIT, rcl_node_type_description_service_init(this->node_ptr));
+}
+
+/* Basic nominal test of the ~/get_type_description service. */
+TEST_F(CLASSNAME(TestGetTypeDescSrvEnabledFixture, RMW_IMPLEMENTATION), test_service_nominal) {
+  rcl_ret_t ret;
+  const rosidl_service_type_support_t * ts = ROSIDL_GET_SRV_TYPE_SUPPORT(
+    type_description_interfaces, srv, GetTypeDescription);
+
+  // Create client.
+  rcl_client_t client = rcl_get_zero_initialized_client();
+  rcl_client_options_t client_options = rcl_client_get_default_options();
+  ret = rcl_client_init(
+    &client, this->node_ptr, ts, this->get_type_description_service_name,
+    &client_options);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    rcl_ret_t ret = rcl_client_fini(&client, this->node_ptr);
+    EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  });
+  ASSERT_TRUE(wait_for_server_to_be_available(this->node_ptr, &client, 10, 1000));
+
+  // Initialize a request.
+  type_description_interfaces__srv__GetTypeDescription_Request client_request;
+  type_description_interfaces__srv__GetTypeDescription_Request__init(&client_request);
+
+  // Fill the request. We use the GetTypeDescription hash because we know that that type
+  // is registered.
+  const rosidl_type_hash_t * type_hash = ts->get_type_hash_func(ts);
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  char * type_hash_str;
+  EXPECT_EQ(RCUTILS_RET_OK, rosidl_stringify_type_hash(type_hash, allocator, &type_hash_str));
+  OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+  {
+    allocator.deallocate(type_hash_str, allocator.state);
+  });
+  rosidl_runtime_c__String__assign(&client_request.type_hash, type_hash_str);
+  rosidl_runtime_c__String__assign(&client_request.type_name, GET_TYPE_DESCRIPTION_SRV_TYPE_NAME);
+  client_request.include_type_sources = false;
+
+  // Send the request.
+  int64_t sequence_number;
+  ret = rcl_send_request(&client, &client_request, &sequence_number);
+  type_description_interfaces__srv__GetTypeDescription_Request__fini(&client_request);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  // Initialize the response owned by the client and take the response.
+  type_description_interfaces__srv__GetTypeDescription_Response client_response;
+  type_description_interfaces__srv__GetTypeDescription_Response__init(&client_response);
+
+  // Retrieve the response.
+  rmw_service_info_t header;
+  ret = rcl_take_response_with_info(&client, &header, &client_response);
```

Comment:
Here - I think we need to wait for the response to be ready before taking it. In Connext, the response isn't ready yet at the time this is called. However, if I add a `wait_for_client_to_be_ready` there is some deadlock inside the Connext binaries (non-public code from the deb) so it hangs indefinitely, even with the timeout set. Here's the stack trace where it is hanging:

```
#0  futex_wait (private=0, expected=2, futex_word=0x55555596e938) at ../sysdeps/nptl/futex-internal.h:146
#1  __GI___lll_lock_wait (futex=futex@entry=0x55555596e938, private=0) at ./nptl/lowlevellock.c:49
#2  0x00007ffff7246082 in lll_mutex_lock_optimized (mutex=0x55555596e938) at ./nptl/pthread_mutex_lock.c:48
#3  ___pthread_mutex_lock (mutex=0x55555596e938) at ./nptl/pthread_mutex_lock.c:93
#4  0x00007ffff4ed1139 in RTIOsapiSemaphore_take () from /opt/rti.com/rti_connext_dds-6.0.1/lib/x64Linux4gcc7.3.0/libnddscore.so
#5  0x00007ffff4eb1ace in REDAWorker_enterExclusiveArea () from /opt/rti.com/rti_connext_dds-6.0.1/lib/x64Linux4gcc7.3.0/libnddscore.so
#6  0x00007ffff4eb4465 in REDACursor_modifyReadWriteArea () from /opt/rti.com/rti_connext_dds-6.0.1/lib/x64Linux4gcc7.3.0/libnddscore.so
#7  0x00007ffff4bf8d21 in PRESPsReader_readOrTakeInstanceI () from /opt/rti.com/rti_connext_dds-6.0.1/lib/x64Linux4gcc7.3.0/libnddscore.so
#8  0x00007ffff4bf982c in PRESPsReader_takeInstance () from /opt/rti.com/rti_connext_dds-6.0.1/lib/x64Linux4gcc7.3.0/libnddscore.so
#9  0x00007ffff564c4d3 in DDS_DataReader_read_or_take_instance_untypedI () from /opt/rti.com/rti_connext_dds-6.0.1/lib/x64Linux4gcc7.3.0/libnddsc.so
#10 0x00007ffff5de43dd in rmw_connextdds_take_samples(RMW_Connext_Subscriber*) () from /ws/install/r/rmw_connextdds_common/lib/librmw_connextdds_common_pro.so
#11 0x00007ffff5dabcf8 in RMW_Connext_Subscriber::loan_messages(bool) () from /ws/install/r/rmw_connextdds_common/lib/librmw_connextdds_common_pro.so
#12 0x00007ffff5dbbf97 in RMW_Connext_Subscriber::has_data() () from /ws/install/r/rmw_connextdds_common/lib/librmw_connextdds_common_pro.so
#13 0x00007ffff5db7f42 in RMW_Connext_WaitSet::detach(rmw_subscriptions_s*, rmw_guard_conditions_s*, rmw_services_s*, rmw_clients_s*, rmw_events_s*, unsigned long&) ()
   from /ws/install/r/rmw_connextdds_common/lib/librmw_connextdds_common_pro.so
#14 0x00007ffff5db8f55 in RMW_Connext_WaitSet::wait(rmw_subscriptions_s*, rmw_guard_conditions_s*, rmw_services_s*, rmw_clients_s*, rmw_events_s*, rmw_time_s const*) ()
   from /ws/install/r/rmw_connextdds_common/lib/librmw_connextdds_common_pro.so
#15 0x00007ffff7f5753e in rcl_wait () from /ws/install/r/rcl/lib/librcl.so
#16 0x00005555555a3fd0 in wait_for_client_to_be_ready(rcl_client_s*, rcl_context_s*, unsigned long, long) ()
#17 0x0000555555568921 in TestGetTypeDescSrvEnabledFixture__rmw_connextdds_test_service_nominal_Test::TestBody() ()
```

Note, this doesn't happen with the other two implementations, this test works on those with and without the `wait`

Reply:
More context: the `rcl_node_get_type_description_service_callback` does get called (from a Connext-created thread), but hangs on attempt to actually get the request message. Trying to find what's creating the deadlock

```
Thread 9 (Thread 0x7ffff1a4f640 (LWP 404837) "rR02000a59fd640"):                                                                                                                                                                                              
#0  futex_wait (private=0, expected=2, futex_word=0x555555f53738) at ../sysdeps/nptl/futex-internal.h:146                                                                                                                                                     
#1  __GI___lll_lock_wait (futex=futex@entry=0x555555f53738, private=0) at ./nptl/lowlevellock.c:49                                                                                                                                                            
#2  0x00007ffff7246082 in lll_mutex_lock_optimized (mutex=0x555555f53738) at ./nptl/pthread_mutex_lock.c:48                                                                                                                                                   
#3  ___pthread_mutex_lock (mutex=0x555555f53738) at ./nptl/pthread_mutex_lock.c:93                                                                                                                                                                            
#4  0x00007ffff71a39c3 in __gthread_mutex_lock(pthread_mutex_t*) () from /ws/install/r/rcl_logging_spdlog/lib/librcl_logging_spdlog.so                                                                                                                        
#5  0x00007ffff71a3a7e in std::mutex::lock() () from /ws/install/r/rcl_logging_spdlog/lib/librcl_logging_spdlog.so                                                                                                                                            
#6  0x00007ffff71a48cc in std::lock_guard<std::mutex>::lock_guard(std::mutex&) () from /ws/install/r/rcl_logging_spdlog/lib/librcl_logging_spdlog.so                                                                                                          
#7  0x00007ffff5d67150 in RMW_Connext_Condition::update_state<RMW_Connext_SubscriberStatusCondition::set_data_available(bool)::{lambda()#1}>(RMW_Connext_SubscriberStatusCondition::set_data_available(bool)::{lambda()#1}&&, bool) (this=0x555555f53730, upda
te_condition=..., notify=true) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/include/rmw_connextdds/rmw_waitset_std.hpp:110                                                                                                                            
#8  0x00007ffff5d65d63 in RMW_Connext_SubscriberStatusCondition::set_data_available (this=0x555555f53730, available=true) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/include/rmw_connextdds/rmw_waitset_std.hpp:709                                 
#9  0x00007ffff5d5bf8d in RMW_Connext_Subscriber::loan_messages (this=0x555555f536c0, update_condition=true) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/src/common/rmw_impl.cpp:1621                                                                
#10 0x00007ffff5d66699 in RMW_Connext_Subscriber::loan_messages_if_needed (this=0x555555f536c0, update_condition=true) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/include/rmw_connextdds/rmw_impl.hpp:397                                           
#11 0x00007ffff5d5c161 in RMW_Connext_Subscriber::take_next (this=0x555555f536c0, ros_messages=0x7ffff1a468d0, message_infos=0x7ffff1a46950, max_samples=1, taken=0x7ffff1a468c8, serialized=false, request_writer_handle=0x0) at /ws/src/ros2/rmw_connextdds/
rmw_connextdds_common/src/common/rmw_impl.cpp:1692                                                                                                                                                                                                            
#12 0x00007ffff5d5bd0d in RMW_Connext_Subscriber::take_message (this=0x555555f536c0, ros_message=0x7ffff1a46920, message_info=0x7ffff1a46950, taken=0x7ffff1a4691b, request_writer_handle=0x0) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/src/common
/rmw_impl.cpp:1522                                                                                                                                                                                                                                            
#13 0x00007ffff5d62c0c in RMW_Connext_Service::take_request (this=0x5555556424e0, request_header=0x7ffff1a48330, ros_request=0x7ffff1a48550, taken=0x7ffff1a46a5f) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/src/common/rmw_impl.cpp:2943          
#14 0x00007ffff5d7c6fa in rmw_api_connextdds_take_request (service=0x555555f419f0, request_header=0x7ffff1a48330, ros_request=0x7ffff1a48550, taken=0x7ffff1a46a5f) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/src/common/rmw_service.cpp:69        
#15 0x00007ffff7eec577 in rmw_take_request (service=0x555555f419f0, request_header=0x7ffff1a48330, ros_request=0x7ffff1a48550, taken=0x7ffff1a46a5f) at /ws/src/ros2/rmw_connextdds/rmw_connextdds/src/rmw_api_impl_ndds.cpp:544                              
#16 0x00007ffff7f47d52 in rcl_take_request_with_info (service=0x5555555e7168, request_header=0x7ffff1a48330, ros_request=0x7ffff1a48550) at /ws/src/ros2/rcl/rcl/src/rcl/service.c:344                                                                        
#17 0x00007ffff7f4803d in rcl_take_request (service=0x5555555e7168, request_header=0x7ffff1a48630, ros_request=0x7ffff1a48550) at /ws/src/ros2/rcl/rcl/src/rcl/service.c:381                                                                                  
#18 0x00007ffff7f3f94d in rcl_node_get_type_description_service_callback (user_data=0x55555566abd0, number_of_events=1) at /ws/src/ros2/rcl/rcl/src/rcl/node.c:601                                                                                            
#19 0x00007ffff5d6be46 in operator() (__closure=0x7ffff1a4cf70) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/src/common/rmw_impl_waitset_std.cpp:880                                                                                                  
--Type <RET> for more, q to quit, c to continue without paging--                                                                                                                                                                                              
#20 0x00007ffff5d6de0f in RMW_Connext_Condition::perform_action_and_update_state<RMW_Connext_SubscriberStatusCondition::notify_new_data()::<lambda()>, RMW_Connext_SubscriberStatusCondition::notify_new_data()::<lambda()> >(struct {...} &&, struct {...} &&
) (this=0x555555f53730, action=..., update_condition=...) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/include/rmw_connextdds/rmw_waitset_std.hpp:136                                                                                                 
#21 0x00007ffff5d6bef1 in RMW_Connext_SubscriberStatusCondition::notify_new_data (this=0x555555f53730) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/src/common/rmw_impl_waitset_std.cpp:868                                                           
#22 0x00007ffff5d679f8 in RMW_Connext_DataReaderListener_on_data_available (listener_data=0x555555f53730, reader=0x55555591d6e0) at /ws/src/ros2/rmw_connextdds/rmw_connextdds_common/src/common/rmw_impl_waitset_std.cpp:138                                 
#23 0x00007ffff54d36c3 in DDS_DataReaderListener_forward_onDataAvailable (listener=0x7ffff1a4d040, presReader=0x55555591d4b8, worker=0x5555559e1dc0) at DataReaderListener.c:363                                                                              
#24 0x00007ffff54e845e in DDS_DataReader_impl_forward_onDataAvailable (presListener=0x7ffff5b6f900 <DDS_DataReader_PRESENTATION_LISTENER>, reader=0x55555591d4b8, worker=0x5555559e1dc0) at DataReader.c:6870                                                 
#25 0x00007ffff48eeb84 in PRESPsService_readerNotifyOfReaderQueueChanges (service=0x5555557a4950, committalInfoList=0x555555f691d0, workerStat=0x0, timestamp=0x7ffff1a4d6c8, receptionTimestamp=0x7ffff1a4d6c8, notifySampleRejected=1, topicQuery=0, worker=
0x5555559e1dc0) at PsServiceImpl.c:2167                                                                                                                                                                                                                       
#26 0x00007ffff48f19ab in PRESPsService_readerSampleListenerOnNewData (listener=0x5555557a5048, firstRelevantSn=0x555555a4aee0, nextRelevantRangeStartSn=0x7ffff1a4d6a8, data=0x7ffff1a4da30, reservedCount=1, timestamp=0x7ffff1a4d6c8, storage=0x555555a4ae6
0, worker=0x5555559e1dc0) at PsServiceImpl.c:2893                                                                                                                                                                                                             
#27 0x00007ffff4accc50 in COMMENDSrReaderService_onSubmessage (listener=0x555555a43990, context=0x7fffd4000c30, timestamp=0x7ffff1a4d6c8, storage=0x55555582cc30, worker=0x5555559e1dc0) at SrReaderService.c:2601                                            
#28 0x00007ffff4b44337 in MIGInterpreter_parse (me=0x55555582bf30, context=0x7fffd4000c30, msg=0x7ffff1a4dd70, worker=0x5555559e1dc0) at Interpreter.c:687                                                                                                    
#29 0x00007ffff4a9607f in COMMENDActiveFacadeReceiver_loop (param=0x5555559d1d60) at ActiveFacade.c:647                                                                                                                                                       
#30 0x00007ffff4d419f3 in RTIOsapiThreadChild_onSpawned (param=0x5555559e3e80) at Thread.c:1726                                                                                                                                                               
#31 0x00007ffff7242b43 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442                                                                                                                                                                   
#32 0x00007ffff72d4a00 in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81                
```

Reply:
OK - that above stack trace actually does have the requisite information. This is a reentrant condition on the same mutex.

`notify_new_data` at frame 21 calls `RMW_ConnextCondition::perform_action_and_update_state` which locks its `mutex_internal`. That calls the registered callback function, which tries to `take_request`, which makes it down to the same Condition which then tries to lock, causing a reentrant deadlock.

What I _don't_ understand, then, is how services work in Connext at all

Reply:
If I change https://github.com/ros2/rmw_connextdds/blob/rolling/rmw_connextdds_common/include/rmw_connextdds/rmw_waitset_std.hpp#L141 to `std::recursive_mutex mutex_internal` then the test starts passing. What I don't know is if this is the correct thing to do.

I think I've identified that there are _no_ tests calling any RMW API within a `rcl_service_set_on_new_request_callback` callback. The only test that calls this API is doing a simple variable increment to check that it was called. The only real service tests are at the `rclcpp` and `rclpy` layers where they are using a totally separate callback framework with their client-level waitsets. `rclcpp` exposes this API, but nothing is testing it, the callback passed to the Service constructor goes through this alternate path.

I _believe_ that it should be totally acceptable to call `rmw_take` during the service callback, as it is not a blocking operation, it just returns an error code if data is not available. So, this looks like a bug in `rmw_connextdds` - unfortunately a variable type issue in a public header :facepalm: so it would cause an ABI break to fix it.

Reply:
> I _believe_ that it should be totally acceptable to call `rmw_take` during the service callback, as it is not a blocking operation, it just returns an error code if data is not available. So, this looks like a bug in `rmw_connextdds` - unfortunately a variable type issue in a public header facepalm so it would cause an ABI break to fix it.

If we think that is the right fix, then let's go ahead and open a pull request for `rmw_connextdds` to make that fix.  We can worry about backporting it to Iron afterwards.  It would also be nice to have a test for this situation, though I think that test would probably want to end up here in `rcl` (or maybe in `rmw`) so we can test it for all RMW implementations.

Reply:
Given that we've totally revamped how this works, I'm going to resolve this conversation.  Feel free to reopen if you think that was the wrong thing to do.